update bluebird instrumentation to account for getNewLibraryCopy

ericmustin opened the pull request:
### What does this PR do?

The Bluebird promise library added a helper function [`getNewLibraryCopy`](http://bluebirdjs.com/docs/api/promise.getnewlibrarycopy.html) in versions [2.11.0](https://github.com/petkaantonov/bluebird/releases/tag/v2.11.0) and [3.4.1](https://github.com/petkaantonov/bluebird/releases/tag/v3.4.1) which: 

> Returns a new independent copy of the Bluebird library.

Usage of this helper method in turn causes the dd-trace-js bluebird instrumentation to get dropped. This PR adds instrumentation to ensure that independant copy of the bluebird library is also instrumented. 

### Motivation

This helper method is used by the popular Node ORM Sequelize, seen [ here](https://github.com/sequelize/sequelize/blob/f3110e32b5b35390e956158339cd9e1d2c67f007/lib/promise.js#L3). As a result, if a basic `express/sequelize/(pg|mysql)` MVC application receives concurrent requests, dd-trace-js instrumentation attaches the database spans to the incorrect trace. You can see a quick screenshot of this behavior, and I'm including a link to a basic express/sequelize application and a script for a trivial reproduction of the issue below. Sorry the repro is a bit hacky, it's quite easy to reproduce though if you want to test however you prefer:

Reproduction of previous behavior: 
1. Run express/sequelize application: `https://github.com/ericmustin/sample_sequelize_express_app`
2. with the application listening on `localhost:3000` run test script for making concurrent requests: 
`https://gist.github.com/ericmustin/d70dfc9fdfb78732302a119793ad7c44`

3. Output:
![all db spans on one trace incorrectly](https://cl.ly/a8156952b88b/download/Image%2525202019-12-26%252520at%2525203.17.13%252520PM.png)

And, with this update applied this is handled correctly:

![updated bluebird instrumentation flamegraph](https://d2ddoduugvun08.cloudfront.net/items/393w0f451X2C0k3s1P2V/Image%202019-12-26%20at%203.50.23%20PM.png?X-CloudApp-Visitor-Id=3196791)

![updated bluebird instrumentation traces](https://d2ddoduugvun08.cloudfront.net/items/0q0M291F1R0r2z3f382e/Image%202019-12-26%20at%203.50.10%20PM.png?X-CloudApp-Visitor-Id=3196791)

### Plugin Checklist
<!-- Fill this section if adding or updating a plugin. Remove otherwise. -->

- [x ] Unit tests.
- [ ] TypeScript [definitions][1].
- [ ] TypeScript [tests][2].
- [ ] API [documentation][3].
- [ ] CircleCI [jobs/workflows][4].
- [ ] Plugin is [exported][5].

[1]: https://github.com/DataDog/dd-trace-js/blob/master/index.d.ts
[2]: https://github.com/DataDog/dd-trace-js/blob/master/docs/test.ts
[3]: https://github.com/DataDog/dd-trace-js/blob/master/docs/API.md
[4]: https://github.com/DataDog/dd-trace-js/blob/master/.circleci/config.yml
[5]: https://github.com/DataDog/dd-trace-js/blob/master/packages/dd-trace/src/plugins/index.js

### Additional Notes

A few things. First is that this helper method wasn't added for v2 and v3 until 2.11.0 and 3.4.1, respectively (i've linked to the release notes in the previous section). However I wasn't entirely sure how to handle the versioning bit in `datadog-plugin-bluebird/src/index.js` and ~updating the `dd-trace-js/versions`~ (this is auto generated by yarn, ignore) directory. My thinking for the instrumentation was adding additional objects to the module exports ala

```
  ....
  {
    name: 'bluebird',
    versions: ['^3.4.1']
    patch (Promise, tracer, config) {
      if(Promise.getNewLibraryCopy) {
        this.wrap(Promise, 'getNewLibraryCopy', createGetNewLibraryCopyWrap(tracer, config, this))  
      }
    },
    unpatch (Promise) {
      if(Promise.getNewLibraryCopy) {
        this.unwrap(Promise, 'getNewLibraryCopy')
      }
    }
  },
  {
    name: 'bluebird',
    versions: ['^2.11.0']
    patch (Promise, tracer, config) {
      if(Promise.getNewLibraryCopy) {
        this.wrap(Promise, 'getNewLibraryCopy', createGetNewLibraryCopyWrap(tracer, config, this))  
      }
    },
    unpatch (Promise) {
      if(Promise.getNewLibraryCopy) {
        this.unwrap(Promise, 'getNewLibraryCopy')
      }
    }
  }  
```
But I wasn't entirely if these are incremental or not (ie, should I do all the wrapping in here or just the bit that applies to `^2.11.0` / `^3.4.1`? ~And also is there documentation on how those details in the versions folder are generated?~ _(auto generated by yarn, ignore)_

In the interim I've just added some defensive code in both the unit tests and the instrumentation code to ensure that we only attempt to patch when this function exists. Any guidance here on the proper way to do this ~and how to generate the right files in the versions directory ( is there some stuff in the scripts directory I run?)~  would be much appreciated.

Also just generally let me know if my approach here for accounting for this helper function in bluebird is appropriate. Although this is meant to get sequelize working, as I assume it's a relatively widely used helper function I kept all the instrumentation as part of core bluebird and not something sequelize specific. Lmk if that is way off. This update works for me as far as fixing database spans getting assigned to the wrong trace, but not sure if there's a more canonical way to instrument this sort of helper method.

thanks for taking a look!

ericmustin commented:
just to update, I've gone ahead and added what I _think_ is the right way to do versioning for the plugin, by adding this additional object to the module.exports array. And i think i answered my own question on the `versions/` directory which is that yarn handles that bit

```
  {
    name: 'bluebird',
    versions: ['^2.11.0', '^3.4.1'],
    patch: <...patching code>,
    unpatch: <...unpatching code>
  },
...
```

Anyway lmk what you think! cheers

ericmustin commented:
just a quick update, spoke informally about this with @bengl ,  he pointed out that this had originally failed to `unpatch` the library copies generated by `getNewLibraryCopy`. Which presents a bit of a bind as there is no canonical place to store this sort of reference to a dynamically generated library copy for later unpatching. I tried to address this with a hidden property on the original library and the helper methods `addToLibraryCopies` and `unwrapLibraryCopies` to push wrapped copies of the library during patching, and then iterate/unwrap them during unpatching. Let me know what you think here of this approach

bengl commented on packages/datadog-plugin-bluebird/src/index.js:
```diff
@@ -2,25 +2,56 @@
+
+    Object.defineProperty(originalLibPrototype, DD_LIB_COPIES, {
+      enumerable: false,
+      writable: true
```
You can set the value here rather than on line 29 with `value: libraryCopies`.

You should also add `configurable: true`.

`enumerable: false` is the default, so you can omit it if you want. (Your choice.)

ericmustin replied:
makes sense, updated with `configurable` and `value` and using the `enumerable` default

bengl commented on packages/datadog-plugin-bluebird/src/index.js:
```diff
@@ -2,25 +2,56 @@
-      return result
+      const libraryCopy = getNewLibraryCopy.apply(this, arguments)
+      shim.wrap(libraryCopy.prototype, '_then', tx.createWrapThen(tracer, config))
+      addToLibraryCopies(originalLib.prototype, libraryCopy)
```
Is there any particular reason to stash this on the prototype rather than the class/constructor? Just curious.

ericmustin replied:
nope, no good reason, updated to point to the constructor instead

ericmustin commented:
Thanks for the feedback @bengl , makes sense to me, I've cleaned up the bits mentioned in review and added some additional bluebird tests here for the mechanics introduced for tracking library copies. The tests did require a bit of utility code to isolate testing the `unpatch` calls that probably ought to get abstracted into a generalised spec utility at some point, but didn't want the scope of this to get out of hand.

bengl commented on packages/datadog-plugin-bluebird/test/index.spec.js:
```diff
@@ -14,3 +14,79 @@ assertPromise('bluebird', bluebird => {
+          bluebird = require(`../../../versions/bluebird@${version}`).get()
+        })
+
+        describe('patching', () => {
```
For testing purposes, I think we're less concerned with the implementation details (except that they're not hanging around after unpatch). The previously added test already ensures that it's been patched. The `'unpatching'` tests ensure that we've properly unpatched, so we're good with that I think.

ericmustin replied:
k makes sense, as long as the hidden property logic is specified in the unpatching section i think that's fine, updated this `patching` spec removed

ericmustin commented on packages/datadog-plugin-bluebird/test/index.spec.js:
```diff
@@ -3,3 +3,70 @@
 assertPromise('bluebird')
+
+assertPromise('bluebird', bluebird => {
+  // TODO: remove if statement when running tests only for versions ^2.11.0 and ^3.4.1
```
added this filter as function thaat you can pass as an optional arg to `assertPromise`. Used `semver` here since it looks like this is what we use elsewhere, I think this ought to do the trick but lmk if i'm off based here

ericmustin commented on packages/datadog-plugin-bluebird/test/index.spec.js:
```diff
@@ -3,3 +3,70 @@
+          bluebird = require(`../../../versions/bluebird@${version}`).get()
+        })
+
+        describe('unpatching', () => {
```
took a pass at this and had a bit of trouble doing this in a generalised way that would work for all the promise libraries, going to give this another shot with fresh eyes tomorrow.

ericmustin replied:
Ok, thanks again for the feedback @rochdev and @bengl . Regarding this last outstanding point: 

I've refactored this test to a general unpatching test for the non-native promise specs. Hit one small roadbump here which is that one of the non-native promise libraries we instrument, `promise-js` in a recent version decided to just return a native promise if it's detected.

`https://github.com/kevincennis/promise/blob/b58ef67dd4023139d0aad98ccd0cb60d8a4f9eec/src/promise.js#L7-#L13`

So it makes testing patching/unpatching a bit difficult and breaks the test for one of the versions. So i added a little one liner to skip the test in this edge case where the non-native promise library defers to native promises, sort've a punt but it started to feel out of scope for the change I was trying to test and didn't want to end up going down a rabbit hole on this.

thanks again, lmk what you think.

rochdev replied:
Actually, if the library returns the native Promise, it should not be patched in the first place, otherwise we are double-patching it as native is already supported, so it's a bug in the `promise-js` plugin.

Can you try adding your check to `promise-js` instead? It should not affect its tests since the context is always propagated for native.

ericmustin replied:
Ok gotcha that makes sense. So what I've done is updated promise-js a bit to have slightly different plugins patching/unpatchin for versions from `0.0.3 - 0.0.6`, and versions `>=0.0.7` (when they export native promises instead if it exists per `https://github.com/kevincennis/promise/compare/0.0.6...master` ) .

 I think this is necessary for the case where the user is on the latest promises-js version but does not have native promises on their version of node, and so still would need promise-js promises instrumented.  

Additionally i've added the version filter to the promise-js specs so that we only test patching/unpatching for `0.0.3 - 0.0.6` . I suppose eventually it may make sense to have the versionFilter fine grained enough to work on a per test basis but dont think it's needed for now.

lmk what you think!

rochdev commented on packages/datadog-plugin-bluebird/test/index.spec.js:
```diff
@@ -1,5 +1,10 @@
+
+assertPromise('bluebird', bluebird => {
+  return bluebird.getNewLibraryCopy()
+}, version => { semver.intersects(version, '^2.11.0 || ^3.4.1') })
```
Can the function be moved to `assetPromise` and pass only the range instead?

ericmustin replied:
yup, updated, moved the semver require and versionrange filtering function to `assertPromise` and exposed versionRange as an option to pass in. Should be a bit more DRY now, and i actually ended up needing it for promise-js updates so makes sense move to promise specs

guidsen commented:
Any progress on this? Eagerly waiting for this to be implemented. Want to use Tracing in our production environment :D!

rochdev commented on packages/datadog-plugin-bluebird/test/index.spec.js:
```diff
@@ -3,3 +3,5 @@
 
 assertPromise('bluebird')
+
+assertPromise('bluebird', bluebird => { return bluebird.getNewLibraryCopy() }, '^2.11.0 || ^3.4.1')
```
Small nit, but the `return` can be avoided by removing the curly braces.

ericmustin replied:
updated

rochdev commented on packages/datadog-plugin-promise-js/src/index.js:
```diff
@@ -5,7 +5,21 @@ const tx = require('../../dd-trace/src/plugins/util/promise')
   {
     name: 'promise-js',
-    versions: ['>=0.0.3'],
+    versions: ['>=0.0.7'],
```
Not needed since the native promise will always be returned for Node.

ericmustin replied:
Ok, makes sense, removing the multiple version thing and just keeping the defensive check on native promises to avoid patching a second time

rochdev commented on packages/datadog-plugin-promise-js/test/index.spec.js:
```diff
@@ -2,4 +2,4 @@
 const assertPromise = require('../../dd-trace/test/plugins/promise')
 
-assertPromise('promise-js')
+assertPromise('promise-js', null, '0.0.3 - 0.0.6')
```
Version range is not needed since tests should pass even with native.

ericmustin replied:
updated, given new tests this isn't required

rochdev commented on packages/dd-trace/test/plugins/promise.js:
```diff
@@ -12,95 +13,129 @@ module.exports = (name, factory) => {
-        })
-
-        describe('without configuration', () => {
+        if (!versionRange || semver.intersects(version, versionRange)) {
```
Why not just return instead of moving everything inside the if?

ericmustin replied:
indeed. updated

rochdev commented on packages/dd-trace/test/plugins/promise.js:
```diff
@@ -12,95 +13,129 @@ module.exports = (name, factory) => {
+
+                  args[i] = () => {
+                    expect(tracer.scope().active()).to.equal(differentContextSpan)
+                    expect(span).to.not.equal(differentContextSpan)
```
This could fail if the library uses `async_hooks` internally. A safer check would be to check that the behavior is the same before patching and after unpatching.

ericmustin replied:
thanks @rochdev , that makes sense at a high level.  I've updated this to test, what I believe, to be the best way i can measure before / after patching. I had to do a bit of re-structuring of the test suite and it gave me a bit of trouble to structure the test in a sane way and it's still a little bit hacky but I think this ought to be a scalable approach for broader promise libraries. lmk what you think

rochdev commented:
@guidsen The plan is to include it in the next release. There are just a few small requested changes left and then it's good to go.

guidsen commented:
> @guidsen The plan is to include it in the next release. There are just a few small requested changes left and then it's good to go.

Great to hear. Thanks for the effort.

rochdev commented on packages/datadog-plugin-bluebird/test/index.spec.js:
```diff
@@ -3,3 +3,5 @@
 
 assertPromise('bluebird')
+
+assertPromise('bluebird', bluebird => bluebird.getNewLibraryCopy(), '^2.11.0 || ^3.4.1')
```
Just realized that a test is missing for chaining, for example `bluebird.getNewLibraryCopy().getNewLibraryCopy()`. I'm also not sure if the code will handle it right now since the new class returned is not patched.

ericmustin replied:
it wouldn't handle it right now, tho i'm not sure if that use case is really used in practice? I can try to account for that though, will try to push something up before eod

ericmustin replied:
updated, lmk what you think

guidsen commented:
Hey there. Nice that this is being merged. When are you planning to release this?

rochdev commented:
@guidsen Released in 0.17.0. Sorry for the delay!

guidsen commented:
No worries. Thanks for the help guys!
